packages/app/src/components/DBTraceWaterfallChart.tsx

```diff
@@ -22,233 +27,338 @@ type SpanRow = {
   });
 
-  const rowWhere = useRowWhere({ meta: data?.meta, aliasMap });
-
-  const rows = useMemo(
-    () =>
-      data?.meta != null && data.meta.length > 0 // Sometimes meta has not loaded yet
-        ? data?.data?.map(row => {
-            return {
-              ...row,
-              id: rowWhere(omit(row, ['HyperDXEventType'])),
-            };
-          })
-        : undefined,
-    [data, rowWhere],
-  );
+  const isFetching = traceIsFetching || logIsFetching;
+  const rows = [...traceRowsData, ...logRowsData];
+
+  rows.sort((a, b) => {
+    const aDate = TimestampNano.fromString(a.Timestamp);
+    const bDate = TimestampNano.fromString(b.Timestamp);
+    const secDiff = aDate.getTimeT() - bDate.getTimeT();
+    if (secDiff === 0) {
+      return aDate.getNano() - bDate.getNano();
+    } else {
+      return secDiff;
+    }
+  });
 
   // 3 Edge-cases
   // 1. No spans, just logs (ex. sampling)
   // 2. Spans, but with missing spans inbetween (ex. missing intermediary spans)
   // 3. Spans, with multiple root nodes (ex. somehow disjoint traces fe/be)
 
-  const spanIds = useMemo(() => {
-    return new Set(
-      rows
-        ?.filter(result => result.HyperDXEventType === 'span')
-        .map(result => result.SpanId) ?? [],
-    );
-  }, [rows]);
-
   // Parse out a DAG of spans
-  type Node = SpanRow & { id: string; children: SpanRow[] };
+  type Node = SpanRow & { id: string; parentId: string; children: SpanRow[] };
   const rootNodes: Node[] = [];
-  const nodes: { [SpanId: string]: any } = {};
+  const parentSpanIdsMap = new Map();
+  const nodesMap = new Map();
+  for (const { ParentSpanId, SpanId } of rows ?? []) {
+    if (ParentSpanId && SpanId) parentSpanIdsMap.set(SpanId, ParentSpanId);
+  }
+  let logSpanCount = -1;
   for (const result of rows ?? []) {
+    const { type, HyperDXEventType, SpanId } = result;
+    // ignore everything without spanId
+    if (!SpanId) continue;
+    if (type === SourceKind.Log) logSpanCount += 1;
+
+    // log have dupelicate span id, tag it with -log-couunt
+    const nodeSpanId =
+      type === SourceKind.Log ? `${SpanId}-log-${logSpanCount}` : SpanId;
+    const nodeParentSpanId =
+      type === SourceKind.Log ? SpanId : parentSpanIdsMap.get(SpanId) || '';
+
     const curNode = {
       ...result,
       children: [],
       // In case we were created already previously, inherit the children built so far
-      ...(result.HyperDXEventType === 'span' ? nodes[result.SpanId] : {}),
+      ...(result.HyperDXEventType === 'span' ? nodesMap.get(nodeSpanId) : {}),
     };
-    if (result.HyperDXEventType === 'span') {
-      nodes[result.SpanId] = curNode;
+    if (!nodesMap.has(nodeSpanId)) {
+      nodesMap.set(nodeSpanId, curNode);
     }
 
-    if (
-      result.HyperDXEventType === 'span' &&
-      // If there's no parent defined, or if the parent doesn't exist, we're a root
-      (result.ParentSpanId === '' || !spanIds.has(result.ParentSpanId))
-    ) {
+    const isRootNode =
```

Comment:
I believe this can be a lot more readable if we do the direct boolean expression which I believe is:
```
const isRootNode = HyperDXEventType === 'span' && type !== SourceKind.Log && !nodeParentSpanId;
```

Reply:
oh cool, good call on this. Also since I am removing `HyperDXEventType`, this will become:
```
// root if is trace event and has no parent
  const isRootNode = type === SourceKind.Trace && !nodeParentSpanId;
```